File: app/src/main/java/view/pager/recycler/ViewAdapter.java
package view.pager.recycler;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import view.pager.recycler.model.Data;
import view.pager.recycler.utils.Check;
import view.pager.recycler.viewholder.PagerItemHolder;
import view.pager.recycler.viewholder.TextItemHolder;

/**
 * Created by harneev on 20/09/16.
 */

public class ViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    public static final int VIEW_TYPE_TEXT = 51;
    public final static int VIEW_TYPE_PAGER = 52;

    private Context mContext;

    private List<Data> mDataList = new ArrayList<>();

    public HashMap<Integer, Integer> mViewPageStates = new HashMap<>();

    ViewAdapter(List<Data> list, Context context) {

        if (list != null && list.size() > 0)
            mDataList.addAll(list);

        mContext = context;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        RecyclerView.ViewHolder viewHolder = null;
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        switch (viewType) {

            case VIEW_TYPE_TEXT:
                View userView = inflater.inflate(R.layout.item_recycler_text, parent, false);
                viewHolder = new TextItemHolder(userView);
                break;

            case VIEW_TYPE_PAGER:
                View blockbusterView = inflater.inflate(R.layout.item_recycler_pager, parent, false);
                viewHolder = new PagerItemHolder(blockbusterView);
                break;
        }

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {

        switch (holder.getItemViewType()) {

            case VIEW_TYPE_TEXT:
                TextItemHolder textHolder = (TextItemHolder) holder;
                configureTextItem(textHolder, position);
                break;

            case VIEW_TYPE_PAGER:
                PagerItemHolder pagerHolder = (PagerItemHolder) holder;
                configurePagerHolder(pagerHolder, position);
                break;
        }
    }

    private void configureTextItem(TextItemHolder holder, int position) {

        Data data = mDataList.get(position);

        if (!Check.isEmpty(data.getTextItem()))
            holder.tvTitle.setText(data.getTextItem());
    }

    private void configurePagerHolder(PagerItemHolder holder, int position) {

        Data data = mDataList.get(position);

        CustomPagerAdapter adapter = new CustomPagerAdapter(data.getPagerItemList(), mContext);
        holder.viewPager.setAdapter(adapter);

        if (mViewPageStates.containsKey(position))
            holder.viewPager.setCurrentItem(mViewPageStates.get(position));
    }

    @Override
    public void onViewRecycled(RecyclerView.ViewHolder holder) {

        if (holder instanceof PagerItemHolder) {

            PagerItemHolder viewHolder = (PagerItemHolder) holder;

            mViewPageStates.put(holder.getAdapterPosition(), viewHolder.viewPager.getCurrentItem());
            super.onViewRecycled(holder);
        }
    }

    @Override
    public int getItemViewType(int position) {

        return mDataList.get(position).getViewType();
    }

    @Override
    public int getItemCount() {

        return mDataList.size();
    }
}

File: app/src/main/java/view/pager/recycler/view/ViewAdapter.kt
package view.pager.recycler.view

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

import java.util.ArrayList
import java.util.HashMap

import view.pager.recycler.R
import view.pager.recycler.model.Data
import view.pager.recycler.utils.Check
import view.pager.recycler.view.viewholder.PagerItemHolder
import view.pager.recycler.view.viewholder.TextItemHolder


class ViewAdapter(val context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val mDataList = ArrayList<Data>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        var viewHolder: RecyclerView.ViewHolder? = null
        val inflater = LayoutInflater.from(parent.context)

        when (viewType) {

            VIEW_TYPE_TEXT -> {
                val userView = inflater.inflate(R.layout.item_recycler_text, parent, false)
                viewHolder = TextItemHolder(userView)
            }

            VIEW_TYPE_PAGER -> {
                val blockbusterView = inflater.inflate(R.layout.item_recycler_pager, parent, false)
                viewHolder = PagerItemHolder(blockbusterView)
            }
        }

        return viewHolder!!
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        when (holder.itemViewType) {

            VIEW_TYPE_TEXT -> {
                val textHolder = holder as TextItemHolder
                configureTextItem(textHolder, position)
            }

            VIEW_TYPE_PAGER -> {
                val pagerHolder = holder as PagerItemHolder
                configurePagerHolder(pagerHolder, position)
            }
        }
    }

    private fun configureTextItem(holder: TextItemHolder, position: Int) {

        val data = mDataList[position]

        if (!Check.isEmpty(data.textItem))
            holder.tvTitle.text = data.textItem
    }

    private fun configurePagerHolder(holder: PagerItemHolder, position: Int) {

        val data = mDataList[position]

        val adapter = CustomPagerAdapter(data.pagerItemList, context)
        holder.viewPager.adapter = adapter

        if (mViewPageStates.containsKey(position)) {
            holder.viewPager.currentItem = mViewPageStates[position]!!
        }
    }

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {

        if (holder is PagerItemHolder) {

            mViewPageStates[holder.getAdapterPosition()] = holder.viewPager.currentItem
            super.onViewRecycled(holder)
        }
    }

    override fun getItemViewType(position: Int): Int {

        return mDataList[position].viewType
    }

    override fun getItemCount(): Int {

        return mDataList.size
    }

    companion object {

        internal val VIEW_TYPE_TEXT = 51
        internal val VIEW_TYPE_PAGER = 52
    }
}
